src/util/psbt/mod.rs

```diff
@@ -16,6 +21,138 @@ pub mod serialize;
 mod psbtmap;
 pub use self::psbtmap::{PsbtMap, PsbtGlobal, PsbtInput, PsbtOutput};
 
+/// A Partially Signed Transaction.
+#[derive(Debug, Clone)]
+pub struct PartiallySignedTransaction {
+    /// The key-value pairs for all global data.
+    global: PsbtGlobal,
+    /// The corresponding key-value map for each input in the unsigned
+    /// transaction.
+    inputs: Vec<PsbtInput>,
+    /// The corresponding key-value map for each output in the unsigned
+    /// transaction.
+    outputs: Vec<PsbtOutput>,
+}
+
+impl PartiallySignedTransaction {
+    /// Create a PartiallySignedTransaction from an unsigned transaction, error
+    /// if not unsigned
+    pub fn from_unsigned_tx(tx: Transaction) -> Result<Self, util::Error> {
+        Ok(PartiallySignedTransaction {
+            inputs: Vec::with_capacity(tx.input.len()),
+            outputs: Vec::with_capacity(tx.output.len()),
+            global: Result::from(tx)?,
+        })
+    }
+
+    /// Extract a Transaction from a finalized PartiallySignedTransaction
+    pub fn extract_tx(self) -> Transaction {
+        let mut tx: Transaction = self.global.unsigned_tx;
+
+        for (vin, psbtin) in tx.input.iter_mut().zip(self.inputs.into_iter()) {
+            vin.script_sig = psbtin.final_script_sig.unwrap();
+            vin.witness = psbtin.final_script_witness.unwrap();
+        }
+
+        tx
+    }
+
+    /// Attempt to merge with another `PartiallySignedTransaction`.
+    pub fn merge(&mut self, other: Self) -> Result<(), util::Error> {
+        self.global.merge(other.global)?;
+
+        for (self_input, other_input) in self.inputs.iter_mut().zip(other.inputs.into_iter()) {
+            self_input.merge(other_input)?;
+        }
+
+        for (self_output, other_output) in self.outputs.iter_mut().zip(other.outputs.into_iter()) {
+            self_output.merge(other_output)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<S: SimpleEncoder<Error = util::Error>> ConsensusEncodable<S> for PartiallySignedTransaction {
+    fn consensus_encode(&self, s: &mut S) -> Result<(), S::Error> {
+        [b'p', b's', b'b', b't'].consensus_encode(s)?;
+
+        0xff_u8.consensus_encode(s)?;
+
+        self.global.consensus_encode(s)?;
+
+        for i in &self.inputs {
+            i.consensus_encode(s)?;
+        }
+
+        for i in &self.outputs {
+            i.consensus_encode(s)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<D: SimpleDecoder<Error = util::Error>> ConsensusDecodable<D> for PartiallySignedTransaction {
+    fn consensus_decode(d: &mut D) -> Result<Self, D::Error> {
+        let magic: [u8; 4] = ConsensusDecodable::consensus_decode(d)?;
+
+        if [b'p', b's', b'b', b't'] != magic {
```

Comment:
nit: you can use `*b"psbt"` rather than an explicit array here

Reply:
Awesome! Much cleaner...